### PR TITLE
Add dev sync

### PR DESCRIPTION
**Pull request by vmrob**
Add a dev sync to assist in development mode when a needs file changes a dev mode-enabled library.

### CODE REVIEW

**Comment by vmrob** [comment]
Example from test output:

```
$ needy dev sync
['git', 'fetch']
['git', 'checkout', u'4af956a3936965d7b61afce08092b400ca75b865']
Previous HEAD position was 370bf72... miniupnpc.h changed version to 1.8
HEAD is now at 4af956a... update CMakeLists.txt to build upnpc-static/upnpc-shared executables
['git', 'submodule', 'update', '--init', '--recursive']
```

**Comment by ccbrown** [comment]
I'll look at the code tomorrow, but I don't think I really like all that Git output. In the general case, i would probably prefer a nice `source modifications re-applied to 370bf72e7282752e04e54ee9e5f13a80a95082f3` log message. So maybe the low-level source-dependent commands are something for verbose logging?

File: tests/functional/functional_test.py [hunk 1]
```diff
@@ -16,7 +16,15 @@ def setUp(self):
...
-        shutil.rmtree(self.__path)
+        def rmtree_onerror(func, path, exc_info):
+            ''' from http://stackoverflow.com/questions/2656322/shutil-rmtree-fails-on-windows-with-access-is-denied '''
+            import stat
+            if not os.access(path, os.W_OK):
+                os.chmod(path, stat.S_IWUSR)
+                func(path)
+            else:
+                raise
+        shutil.rmtree(self.__path, onerror=rmtree_onerror)
```
Comment (ccbrown) [inline tests/functional/functional_test.py]: can you move this to `filesystem`? maybe as a `force_rmtree` function? definitely gonna have some use-cases for this soon

File: needy/needy.py [hunk 2]
```diff
@@ -428,6 +428,30 @@ def initialize(self, target, filters=None):
             print('Initializing {}...'.format(name))
             libraries[0].initialize_source()
 
+    def synchronize(self, target, filters=None):
+        needs_configuration = self.needs_configuration(target)
```
Comment (ccbrown) [inline needy/needy.py]: doesn't look like you use this variable

File: needy/needy.py [hunk 3]
```diff
@@ -428,6 +428,30 @@ def initialize(self, target, filters=None):
...
+                raise RuntimeError('{} does not have development mode enabled'.format(name))
+            elif name in dev_mode_libraries:
+                libraries_to_sync.append((name, libraries))
+
+        if len(libraries_to_sync) == 0:
+            logging.warning('no libraries have development mode enabled')
+
+        for name, libraries in libraries_to_sync:
+            assert len(libraries) == 1
+            print('Synchronizing {}...'.format(name))
```
Comment (ccbrown) [inline needy/needy.py]: should use `logging`

File: needy/library.py [hunk 4]
```diff
@@ -105,6 +105,18 @@ def initialize_source(self):
...
 
+    def synchronize_source(self):
+        if 'download' in self.__configuration:
+            source = Download(self.__configuration['download'], self.__configuration['checksum'], self.source_directory(), os.path.join(self.directory(), 'download'))
+        elif 'repository' in self.__configuration:
+            source = GitRepository(self.__configuration['repository'], self.__configuration['commit'], self.source_directory())
+        elif 'directory' in self.__configuration:
+            source = Directory(self.__configuration['directory'] if os.path.isabs(self.__configuration['directory']) else os.path.join(self.needy.path(), self.__configuration['directory']), self.source_directory())
+        else:
+            raise ValueError('no source specified in configuration')
```
Comment (ccbrown) [inline needy/library.py]: hmmmmmmmm... these 9 lines look vaguely familiar...

i think you know what needs to be done here.

File: needy/needy.py (hunk 2 above)
Comment (vmrob) [inline needy/needy.py]: The line below. I'll coalesce them.

**Comment by vmrob** [comment]
To reiterate what we discussed in person:

- Good idea to remove the git-specific messages if/when other sources are able to utilize this
- Good idea to ensure that errors always show the complete message.

**Pull request merged**